cmd/konvoy-image/cmd/vsphere.go

```diff
@@ -0,0 +1,139 @@
+package cmd
+
+import (
+	"github.com/spf13/cobra"
+	flag "github.com/spf13/pflag"
+
+	"github.com/mesosphere/konvoy-image-builder/pkg/app"
+)
+
+var (
+	vsphereExample = "vsphere --datacenter dc1 --cluster zone1 --datastore nfs-store1 " +
+		"--network public --template=d2iq-base-templates/d2iq-base-CentOS-7.9 " +
+		"images/ami/centos-79.yaml"
+	vsphereUse = "vsphere <image.yaml>"
+)
+
+func NewVSphereBuildCmd() *cobra.Command {
+	flags := &buildCLIFlags{}
+	cmd := &cobra.Command{
+		Use:     vsphereUse,
+		Short:   "build and provision vsphere images",
+		Example: vsphereExample,
+		Args:    cobra.ExactArgs(1),
+		Run: func(cmd *cobra.Command, args []string) {
+			runBuild(args[0], flags)
+		},
+	}
+
+	initBuildVSphereFlags(cmd.Flags(), flags)
+	return cmd
+}
+
+func NewVSphereGenerateCmd() *cobra.Command {
+	flags := &generateCLIFlags{}
+	cmd := &cobra.Command{
+		Use:     vsphereUse,
+		Short:   "generate files relating to building vsphere images",
+		Example: vsphereExample,
+		Args:    cobra.ExactArgs(1),
+		Run: func(cmd *cobra.Command, args []string) {
+			runGenerate(args[0], flags)
+		},
+	}
+
+	initGenerateVSphereFlags(cmd.Flags(), flags)
+	return cmd
+}
+
+func initBuildVSphereFlags(fs *flag.FlagSet, buildFlags *buildCLIFlags) {
+	initGenerateArgs(fs, &buildFlags.generateCLIFlags)
+	initVSphereArgs(fs, &buildFlags.generateCLIFlags)
+
+	addBuildArgs(fs, buildFlags)
+}
+
+func initGenerateVSphereFlags(fs *flag.FlagSet, generateFlags *generateCLIFlags) {
+	initGenerateArgs(fs, generateFlags)
+	initVSphereArgs(fs, generateFlags)
+}
+
+func initVSphereArgs(fs *flag.FlagSet, gFlags *generateCLIFlags) {
+	gFlags.userArgs.VSphere = &app.VSphereArgs{}
+	addVSphereArgs(fs, gFlags.userArgs.VSphere)
+}
+
+func addVSphereArgs(fs *flag.FlagSet, vsphereArgs *app.VSphereArgs) {
+	fs.StringVar(
+		&vsphereArgs.Template,
+		"template",
+		"",
+		"Base template to be used. Can include folder. <templatename> or <folder>/<templatename> (REQUIRED)",
+	)
+	fs.StringVar(
+		&vsphereArgs.Datacenter,
+		"datacenter",
+		"",
+		"vSphere datacenter (REQUIRED)",
```

Comment:
These aren't required 

```
bin/konvoy-image-wrapper with args [build vsphere images/ova/rhel-79.yaml --dry-run --overrides=overrides/offline.yaml --overrides=packer-ova-offline-override.yaml]
```

They can be in the overrides or passed in through these flags.

Reply:
why is datacenter not required? 

I don't think it will run without and if we have some defaults somewhere we rather should delete these defaults as customer resources will definitly be named differently

Reply:
https://github.com/mesosphere/konvoy-image-builder/blob/main/pkg/packer/manifests/vsphere/packer.pkr.hcl#L179-L182C2

we could name the additional variable here. But no matter how its set datacenter is needed

Reply:
> why is datacenter not required?

> I don't think it will run without and if we have some defaults somewhere we rather should delete these defaults as customer resources will definitly be named differently

we do not have defaults anywhere. this is being passed in through an override like in the command I posted 

```
--overrides=packer-ova-offline-override.yaml
```

It wouldn't be good to force the user to only use this flag when it was being passed in previously via override.

Therefore, setting the flag is not required. The value of datacenter is required. I hope this clears things up

Reply:
i see... 

Its probably just my personal perspective on these things...

So the `--overrides` as well as environment variable are just another way to full fill the requirement here. Which lets me still think that data is required. 

- So we agree that data center is a required input.
- KIB offers multiple ways to inject this data

Having the sub-command with this PR makes the step of having the override file completely obsolete. 

The huge benefit is its self-documenting.

When we drop the required mark from `--datacenter` we must add additional documentation that now either `--overrides` or env `VSPHERE_DATACENTER` are being required. Which IMO makes it more complicated to understand.

> It wouldn't be good to force the user to only use this flag when it was being passed in previously via override
This sub command was not existent so we do not force anything as plain `build` stays absolutely the same

Reply:
> When we drop the required mark from --datacenter we must add additional documentation that now either --overrides or env VSPHERE_DATACENTER are being required. Which IMO makes it more complicated to understand.

I think we can add this in the flag description "must be passed in via this flag, override file, or included in ova template yaml file"

Reply:
So I just looked into the other sub-commands. They just completely miss what is required and what is optional.

While neither environment variables as well as the override stuff is documented in the command itself. You have to dig deep into our documentation to understand this features.

While many things in public clouds could be assumed this won't be possible in customer driven environments like vsphere where every single dependent resource is custom defined.


I just tried to add some additional notes into the flags but it just makes things complicated.



So can you tell me why documenting that this flag is required is a problem at all?

Leaving it out makes the documentation worse which is something we should not even consider due to this process already being over-complicated.


Frankly I would even question if the `<image.yaml>` input could be left out. In vsphere it only has use-less information which gets forwarded into ansible while ansible by its nature knows already about it. 
The only information left in that file is IMO not really important as it just controls some metadata for the template

Reply:
Great idea thank you !

Reply:
> I would even question if the <image.yaml> input could be left out
We are moving out <image.yaml> in new design. I DM you details.